File: src/com/saltuk/lru/cache/LRUCache.java
package com.saltuk.lru.cache;

import java.util.Map;
import java.util.HashMap;

/*
* 146. LRU Cache
 * */
class LRUCache {
    private Map<Integer, Info> keyToInfo = new HashMap<>();
    private NodeList nodes = new NodeList();
    private int capacity;

    public LRUCache(int capacity) {
        this.capacity = capacity;
    }

    public int get(int key) {
        var info = keyToInfo.get(key);

        if (info == null) {
            return -1;
        }

        nodes.moveTop(info.node);

        return info.value;
    }

    public void put(int key, int value) {
        if (keyToInfo.containsKey(key)) {
            update(key, value);
        } else {
            add(key, value);
        }
    }

    private void update(int key, int value) {
        var info = keyToInfo.get(key);
        nodes.remove(info.node);
        var newNode = nodes.addNode(key);
        keyToInfo.put(key, new Info(value, newNode));
    }

    private void add(int key, int value) {
        if (keyToInfo.size() >= capacity) {
            var nodeToRemove = nodes.getTail();
            nodes.remove(nodeToRemove);
            var removedKey = nodeToRemove.getKey();
            keyToInfo.remove(removedKey);
        }
        var newNode = nodes.addNode(key);
        keyToInfo.put(key, new Info(value, newNode));
    }

    class NodeList {
        private Node head = null;
        private Node tail = null;

        public void moveTop(Node node) {
            var nodeNext = node.getNext();
            var nodePrev = node.getPrev();

            if (nodePrev == null) {
                return;
            }

            if (nodeNext == null) {
                tail = nodePrev;
            } else {
                nodeNext.setPrev(nodePrev);
            }

            nodePrev.setNext(nodeNext);

            node.setNext(head);
            node.setPrev(null);

            head.setPrev(node);
            head = node;
        }

        public void remove(Node node) {
            var nodeNext = node.getNext();
            var nodePrev = node.getPrev();

            if(nodePrev == null) {
                head = nodeNext;
            } else {
                nodePrev.setNext(nodeNext);
            }
            if (nodeNext == null) {
                tail = nodePrev;
            } else {
                nodeNext.setPrev(nodePrev);
            }
        }

        public Node addNode(int key) {
            var node = new Node(key, null, head);
            if (head != null) {
                head.setPrev(node);
            }
            if (tail == null) {
                tail = node;
            }
            head = node;
            return node;
        }

        public Node getTail() {
            return tail;
        }

    }

    record Info(int value, Node node){}

    class Node {
        private final int key;
        private Node prev;
        private Node next;

        public Node(int key, Node prev, Node next) {
            this.key = key;
            this.prev = prev;
            this.next = next;
        }

        public int getKey() {
            return key;
        }

        public Node getPrev() {
            return prev;
        }

        public Node getNext() {
            return next;
        }

        public void setPrev(Node prev) {
            this.prev = prev;
        }

        public void setNext(Node next) {
            this.next = next;
        }
    }
}

/**
 * Your com.saltuk.lru.cache.LRUCache object will be instantiated and called as such:
 * com.saltuk.lru.cache.LRUCache obj = new com.saltuk.lru.cache.LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/src/LRUCache.kt
class LRUCache(private val capacity: Int) {
    private val cache = HashMap<Int, Info<Int, Int>>()
    private val linkedList = LinkedList<Int>()
    private var size = 0

    fun get(key: Int): Int {
        val data = cache[key] ?: return -1

        linkedList.moveToTail(data.node)

        return data.value
    }

    fun put(key: Int, value: Int) {
        if (cache[key] != null) {
            val node = cache[key]!!.node
            linkedList.moveToTail(node)
            cache[key] = Info(value, node)
            return
        }
        if (size >= capacity) {
            val keyToRemove = linkedList.getHeadValue()
            cache.remove(keyToRemove)
            linkedList.moveHeadForward()
            size--
        }
        cache[key] = Info(value, linkedList.addValue(key))
        size++
    }

    data class Info<K, V>(val value: V, val node: LinkedList.Node<K>)
    class LinkedList<T> {
        private var head: Node<T>? = null
        private var tail: Node<T>? = null

        fun moveToTail(node: Node<T>) {
            if (node == tail) {
                return
            }

            if (head == node) {
                head = node.next
            }

            node.prev?.next = node.next
            node.next?.prev = node.prev

            node.prev = tail
            node.next = null

            tail?.next = node
            tail = node
        }

        fun getHeadValue() : T? {
            return head?.value
        }

        fun addValue(value: T): Node<T> {
            val newNode = Node(value, tail, null)
            tail?.next = newNode
            tail = newNode
            if(head == null) {
                head = newNode
            }
            return newNode
        }

        fun moveHeadForward() {
            head = head?.next
        }

        class Node<T>(val value: T, var prev: Node<T>?, var next: Node<T>?)
    }
}